### PR TITLE
Explainability Dataset Task: `GridGraph`

### PR DESCRIPTION
**Pull request by shenoynikhil**
Adds GridGraph. Fixes task https://github.com/pyg-team/pytorch_geometric/issues/5817.

```python
from torch_geometric.datasets.generators import GridGraph

motif = Motif(structure='house')
generator = GridGraph(height=10, width=10, dtype=torch.float)
dataset = ExplainerDataset(generator=generator)
```

TODO
- [x] Tests
- [x] ChangeLog

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@rusty1s Maybe you can have a look at this PR.

File: torch_geometric/datasets/graph_generator/__init__.py [hunk 1]
```diff
@@ -1,9 +1,6 @@
...
 from .ba_graph import BAGraph
 from .er_graph import ERGraph
+from .grid_graph import GridGraph
 
-__all__ = classes = [
-    'GraphGenerator',
-    'BAGraph',
-    'ERGraph',
-]
+__all__ = classes = ['GraphGenerator', 'BAGraph', 'ERGraph', 'GridGraph']
```
Comment (rusty1s) [inline torch_geometric/datasets/graph_generator/__init__.py:6]:
```suggestion
__all__ = classes = ['GraphGenerator', 'BAGraph', 'ERGraph', 'GridGraph', ]
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ## [2.3.0] - 2023-MM-DD
 ### Added
+- Added `GridGraph` graph generator to generate Grid graphs ([#6220](https://github.com/pyg-team/pytorch_geometric/pull/6220)
```
Comment (rusty1s) [inline CHANGELOG.md:8]:
```suggestion
- Added `GridGraph` graph generator to generate grid graphs ([#6220](https://github.com/pyg-team/pytorch_geometric/pull/6220)
```

File: torch_geometric/datasets/graph_generator/grid_graph.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.datasets.graph_generator import GraphGenerator
+from torch_geometric.utils import grid
+
+
+class GridGraph(GraphGenerator):
+    r"""Generates Grid Graphs.
```
Comment (rusty1s) [inline torch_geometric/datasets/graph_generator/grid_graph.py:11]:
```suggestion
    r"""Generates two-dimensional grid graphs.
```

File: torch_geometric/datasets/graph_generator/grid_graph.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    See :meth:`~torch_geometric.utils.grid` for more information.
+
+    Args:
+        height (int): The height of the grid.
+        width (int): The width of the grid.
+        dtype (:obj:`torch.dtype`, optional): The desired data type of the
+            returned position tensor.
```
Comment (rusty1s) [inline torch_geometric/datasets/graph_generator/grid_graph.py:18]:
```suggestion
            returned position tensor. (default: :obj:`None`)
```

**Review by rusty1s — APPROVED** [review]
Thank you, this is great :)

**Pull request merged**